### PR TITLE
Blog post about pagination and mutiny

### PR DESCRIPTION
**Pull request by cescoffier**
A blog post about Mutiny and pagination.

### CODE REVIEW

**Comment by cescoffier** [comment]
CC @jponge.

File: _posts/2020-09-15-mutiny-pagination.adoc [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+synopsis: How to use API using pagination with Mutiny
+author: cescoffier
+---
+
+At the beginning of the Mutiny adventure, my friend Alex came to me with an interesting problem. 
+Alex wanted to retrieve data from a REST service in a reactive manner.
+So far, no problem, we have everything for this in our toolbox. 
+But, this service, as many services, is using pagination. 
+Ah! That makes things a bit more spicy.
+Alex wanted to retrieve all the items and consumes them as a stream, but you can't retrieve the items in one batch.
```
Comment (kenfinnigan) [inline _posts/2020-09-15-mutiny-pagination.adoc:15]: s/consumes/consume ?

File: _posts/2020-09-15-mutiny-pagination.adoc [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+        first beer
+    },
+    {
+        second beer
+    }, 
+    // ...
+]
+----
+
+I won't details the content of each object, the documentation page does a great job about that.
```
Comment (kenfinnigan) [inline _posts/2020-09-15-mutiny-pagination.adoc:43]: Maybe missing "show" or something like that between "won't" and "details?

File: _posts/2020-09-15-mutiny-pagination.adoc [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+Let's focus on the pagination aspect.
+First, we passed the `page` query parameter, indicating which page we want.
+Generally, when you retrieve a page, the API provides a way to know if there is a next page (a special field in the JSON document, or HTTP header), but the Punk API does not provide any hint. 
+So, to retrieve all the beers, we need to invoke the service for page 1, 2, 3... until the returned JSON array is empty. 
+
+In an imperative word, to retrieve all the beers, you would do something like this:
```
Comment (kenfinnigan) [inline _posts/2020-09-15-mutiny-pagination.adoc:49]: s/word/world

File: _posts/2020-09-15-mutiny-pagination.adoc [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+
+## Retrieving multiple page
+
+So, now, we know how to retrieve a single page and extract the items from it. 
+We just need to repeat this operation for every page, and provide a stream.
+
+Mutiny provides a method to create a `Multi` by repeating multiple times a `Uni`.
+Under the hood, it calls a method returning a `Uni` and subscribe on it.
+But we need to make _progress_, and pass the current page.
+Mutiny offers the possibility to store a state, to the method creating the `Uni` can increment and get the page number:
```
Comment (kenfinnigan) [inline _posts/2020-09-15-mutiny-pagination.adoc:115]: I think ", to the method" should be ", so the method"?

File: _posts/2020-09-15-mutiny-pagination.adoc [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+     getPage(client, page.incrementAndGet())
+)
+.until(List::isEmpty);
+----
+
+The `until` clause indicates when the iteration must be stopped. 
+
+## Unpacking the beers
+
+We know have a stream of list of beers.
```
Comment (kenfinnigan) [inline _posts/2020-09-15-mutiny-pagination.adoc:143]: s/know/now

Comment (kenfinnigan) [reply _posts/2020-09-15-mutiny-pagination.adoc:143]: Maybe also reword "stream of list of beers"?

File: _posts/2020-09-15-mutiny-pagination.adoc [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+
+However, at some point, we must stop. 
+As we said earlier, we can stop when the returned list is empty:
+
+[source, java]
+----
+Multi<List<Beer>> multi = Multi.createBy().repeating().uni(AtomicInteger::new, page ->
+     getPage(client, page.incrementAndGet())
+)
+.until(List::isEmpty);
```
Comment (kenfinnigan) [inline _posts/2020-09-15-mutiny-pagination.adoc:136]: For my own understanding, is the `until` waiting for the `List<Beer>` on the `Mutli` to be empty, or the `List` returned from `getPage`?

**Review by kenfinnigan — CHANGES_REQUESTED** [review]

File: _posts/2020-09-15-mutiny-pagination.adoc (hunk 4 above)
Comment (cescoffier) [reply _posts/2020-09-15-mutiny-pagination.adoc:115]: Replacing it with: `Mutiny offers the possibility to store a state in order to let the method creating the `Uni` increments the page number:`

File: _posts/2020-09-15-mutiny-pagination.adoc (hunk 6 above)
Comment (cescoffier) [reply _posts/2020-09-15-mutiny-pagination.adoc:136]: the list from the `getPage()` method, which means that it waits until the received item (the list) is empty.

File: _posts/2020-09-15-mutiny-pagination.adoc (hunk 1 above)
Comment (cescoffier) [reply _posts/2020-09-15-mutiny-pagination.adoc]: Fixed.

File: _posts/2020-09-15-mutiny-pagination.adoc (hunk 3 above)
Comment (cescoffier) [reply _posts/2020-09-15-mutiny-pagination.adoc]: Fixed

File: _posts/2020-09-15-mutiny-pagination.adoc (hunk 2 above)
Comment (cescoffier) [reply _posts/2020-09-15-mutiny-pagination.adoc]: Fixed.

File: _posts/2020-09-15-mutiny-pagination.adoc (hunk 5 above)
Comment (cescoffier) [reply _posts/2020-09-15-mutiny-pagination.adoc]: Can you check if it's better?

File: _posts/2020-09-15-mutiny-pagination.adoc (hunk 6 above)
Comment (cescoffier) [reply _posts/2020-09-15-mutiny-pagination.adoc:136]: Ok, I've added a bit more explanation, can you check?

Comment (kenfinnigan) [reply _posts/2020-09-15-mutiny-pagination.adoc:136]: is that because the `until` is essentially part of the `repeating`?

**Review by kenfinnigan — APPROVED** [review]

File: _posts/2020-09-15-mutiny-pagination.adoc (hunk 6 above)
Comment (cescoffier) [reply _posts/2020-09-15-mutiny-pagination.adoc:136]: yes, it is. `repeating` returns an object to configure when the repetition stops, not a multi. `until` returns the `multi`.

**Pull request merged**